ext/drx/scatter_gather_aarch64.c

```diff
@@ -38,6 +38,479 @@
+        DRX_CASE(ldnt1w, 1, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ldnt1d, 1, OPSZ_8, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ldnt1sb, 1, OPSZ_1, true, DRX_NORMAL_FAULTING);
+        DRX_CASE(ldnt1sh, 1, OPSZ_2, true, DRX_NORMAL_FAULTING);
+        DRX_CASE(ldnt1sw, 1, OPSZ_4, true, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(st1b, 1, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st1h, 1, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st1w, 1, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st1d, 1, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(stnt1b, 1, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(stnt1h, 1, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(stnt1w, 1, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(stnt1d, 1, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(ld2b, 2, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld2h, 2, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld2w, 2, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld2d, 2, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(st2b, 2, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st2h, 2, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st2w, 2, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st2d, 2, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(ld3b, 3, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld3h, 3, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld3w, 3, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld3d, 3, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(st3b, 3, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st3h, 3, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st3w, 3, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st3d, 3, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(ld4b, 4, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld4h, 4, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld4w, 4, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld4d, 4, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(st4b, 4, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st4h, 4, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st4w, 4, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st4d, 4, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(ld1rob, 1, OPSZ_1, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(ld1rqb, 1, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld1rqh, 1, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld1rqw, 1, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld1rqd, 1, OPSZ_8, false, DRX_NORMAL_FAULTING);
+#undef DRX_CASE
+
+    default: DR_ASSERT_MSG(false, "Invalid scatter/gather instruction");
+    }
+}
+
+#define EMIT(op, ...)    \
+    instrlist_preinsert( \
+        bb, sg_instr, INSTR_XL8(INSTR_CREATE_##op(drcontext, __VA_ARGS__), orig_app_pc))
```

Comment:
With these implicit variables it is probably best to move this define inside the function (or else make these actual parameters).